File: CPlusPlus/ExampleTests/test_shopping_cart.cpp
#include <gtest/gtest.h>
#include <iostream>
#include "common/common.h"
#include "shopping_cart_glue.h"

TEST(ShoppingCart, Scenario_AddItems) {
    ShoppingCartGlue glue;
    std::vector<CatalogItemString> objectList1 = {
        CatalogItemString::from_vec({"Widget", "10"}),
        CatalogItemString::from_vec({"WhatCallIt", "20"}),
        CatalogItemString::from_vec({"ThingaMaJig", "30"}),
    };
    glue.given_catalog_has(objectList1);

    std::vector<OrderItemString> objectList2 = {
    };
    glue.given_item_collection_is(objectList2);

    std::vector<OrderItemString> objectList3 = {
        OrderItemString::from_vec({"Widget", "2", "1", "1"}),
    };
    glue.when_item_added(objectList3);

    std::vector<OrderItemString> objectList4 = {
        OrderItemString::from_vec({"Widget", "2", "$10.00", "$20.00"}),
    };
    glue.then_item_collection_is(objectList4);

    std::vector<OrderItemString> objectList5 = {
        OrderItemString::from_vec({"WhatCallIt", "3", "1", "1"}),
    };
    glue.when_item_added(objectList5);

    std::vector<OrderItemString> objectList6 = {
        OrderItemString::from_vec({"Widget", "2", "$10.00", "$20.00"}),
        OrderItemString::from_vec({"WhatCallIt", "3", "$20.00", "$60.00"}),
    };
    glue.then_item_collection_is(objectList6);

    std::vector<ItemPriceInputString> objectList7 = {
        ItemPriceInputString::from_vec({"$80"}),
    };
    glue.then_total_of_items_is(objectList7);

}

TEST(ShoppingCart, Scenario_AShoppingCartWithAddresses) {
    ShoppingCartGlue glue;
    std::vector<CatalogItemString> objectList8 = {
        CatalogItemString::from_vec({"Widget", "10"}),
        CatalogItemString::from_vec({"WhatCallIt", "20"}),
        CatalogItemString::from_vec({"ThingaMaJig", "30"}),
    };
    glue.given_catalog_has(objectList8);

    std::vector<ShoppingCartString> objectList9 = {
        ShoppingCartString{"=EmptyCart", "$0", "$0", "$0", AddressString{"2 Apple Lane", "Somewhere", "NC", "27706"}, AddressString{"1 Apple Lane", "Somewhere", "NC", "27705"}},
    };
    glue.given_shopping_cart(objectList9);

}

TEST(ShoppingCart, Scenario_AddItemsToShoppingCart) {
    ShoppingCartGlue glue;
    std::vector<CatalogItemString> objectList10 = {
        CatalogItemString::from_vec({"Widget", "10"}),
        CatalogItemString::from_vec({"WhatCallIt", "20"}),
        CatalogItemString::from_vec({"ThingaMaJig", "30"}),
    };
    glue.given_catalog_has(objectList10);

    std::vector<ShoppingCartString> objectList11 = {
        ShoppingCartString{"=EmptyCart", "$0", "$0", "$0", AddressString{"", "", "", ""}, AddressString{"", "", "", ""}},
    };
    glue.given_shopping_cart(objectList11);

    std::vector<OrderItemString> objectList12 = {
        OrderItemString::from_vec({"Widget", "2", "1", "1"}),
    };
    glue.when_item_added(objectList12);

    std::vector<OrderItemString> objectList13 = {
        OrderItemString::from_vec({"WhatCallIt", "3", "1", "1"}),
    };
    glue.when_item_added(objectList13);

    std::vector<ShoppingCartString> objectList14 = {
        ShoppingCartString{"=TwoItemCart", "$5", "$4", "$81", AddressString{"", "", "", ""}, AddressString{"", "", "", ""}},
    };
    glue.then_shopping_cart_is(objectList14);

}

TEST(ShoppingCart, Scenario_CostOfEmptyOrderItemCollection) {
    ShoppingCartGlue glue;
    std::vector<CatalogItemString> objectList15 = {
        CatalogItemString::from_vec({"Widget", "10"}),
        CatalogItemString::from_vec({"WhatCallIt", "20"}),
        CatalogItemString::from_vec({"ThingaMaJig", "30"}),
    };
    glue.given_catalog_has(objectList15);

    std::vector<OrderItemString> objectList16 = {
    };
    glue.given_item_collection(objectList16);

    std::vector<ItemPriceInputString> objectList17 = {
        ItemPriceInputString::from_vec({"$0"}),
    };
    glue.then_total_of_items_is(objectList17);

}

TEST(ShoppingCart, BusinessRule_TotalCartPrice) {
    ShoppingCartGlue glue;
    std::vector<CartInputString> objectList18 = {
        CartInputString::from_vec({"$110", "$5", "$11", "$104", "Discount applied before shipping calculated"}),
        CartInputString::from_vec({"$80", "$5", "$4", "$81", ""}),
    };
    glue.examples_businessrule_total_cart_price(objectList18);
}

TEST(ShoppingCart, BusinessRule_ShippingCost) {
    ShoppingCartGlue glue;
    std::vector<ShippingInputString> objectList19 = {
        ShippingInputString::from_vec({"$99.99", "$5.00", "Less than $100"}),
        ShippingInputString::from_vec({"$100.00", "$0", "Free if $100 or more"}),
    };
    glue.examples_businessrule_shipping_cost(objectList19);
}

TEST(ShoppingCart, BusinessRule_Discount) {
    ShoppingCartGlue glue;
    std::vector<DiscountInputString> objectList20 = {
        DiscountInputString::from_vec({"$24.99", "0", ""}),
        DiscountInputString::from_vec({"$25.00", "5", ""}),
        DiscountInputString::from_vec({"$99.99", "5", ""}),
        DiscountInputString::from_vec({"$100.00", "10", ""}),
    };
    glue.examples_businessrule_discount(objectList20);
}

TEST(ShoppingCart, DataType_Percentage) {
    ShoppingCartGlue glue;
    std::vector<ValidValuesString> objectList21 = {
        ValidValuesString::from_vec({"0", "y", ""}),
        ValidValuesString::from_vec({"99", "y", ""}),
        ValidValuesString::from_vec({"100", "y", ""}),
        ValidValuesString::from_vec({"101", "n", ""}),
        ValidValuesString::from_vec({"-1", "n", ""}),
    };
    glue.examples_datatype_percentage(objectList21);
}

File: Rust/ExampleTests/test_shopping_cart.rs
#![allow(unused_mut, unused_variables, unused_imports)]

use crate::common::*;
use super::shopping_cart_glue::ShoppingCartGlue;

// --- Scenario Tests ---

#[test]
fn scenario_add_items() {
    let mut glue = ShoppingCartGlue::new();
    glue.given_catalog_has(&[
        CatalogItemString::from_vec(&["Widget", "10"]),
        CatalogItemString::from_vec(&["WhatCallIt", "20"]),
        CatalogItemString::from_vec(&["ThingaMaJig", "30"]),
    ]);
    glue.given_item_collection_is(&[
    ]);
    glue.when_item_added(&[OrderItemString::from_vec(&["Widget", "2", "1", "1"])]);
    glue.then_item_collection_is(&[OrderItemString::from_vec(&["Widget", "2", "$10.00", "$20.00"])]);
    glue.when_item_added(&[OrderItemString::from_vec(&["WhatCallIt", "3", "1", "1"])]);
    glue.then_item_collection_is(&[
        OrderItemString::from_vec(&["Widget", "2", "$10.00", "$20.00"]),
        OrderItemString::from_vec(&["WhatCallIt", "3", "$20.00", "$60.00"]),
    ]);
    glue.then_total_of_items_is(&[ItemPriceInputString::from_vec(&["$80"])]);
}

#[test]
fn scenario_a_shoppingcart_with_addresses() {
    let mut glue = ShoppingCartGlue::new();
    glue.given_catalog_has(&[
        CatalogItemString::from_vec(&["Widget", "10"]),
        CatalogItemString::from_vec(&["WhatCallIt", "20"]),
        CatalogItemString::from_vec(&["ThingaMaJig", "30"]),
    ]);
    glue.given_shopping_cart(&[ShoppingCartString { items: "=EmptyCart".to_string(), shipping: "$0".to_string(), discount: "$0".to_string(), totalprice: "$0".to_string(), shippingaddress: AddressString { street: "2 Apple Lane".to_string(), city: "Somewhere".to_string(), state: "NC".to_string(), zip: "27706".to_string() }, billingaddress: AddressString { street: "1 Apple Lane".to_string(), city: "Somewhere".to_string(), state: "NC".to_string(), zip: "27705".to_string() } }]);
}

#[test]
fn scenario_add_items_to_shopping_cart() {
    let mut glue = ShoppingCartGlue::new();
    glue.given_catalog_has(&[
        CatalogItemString::from_vec(&["Widget", "10"]),
        CatalogItemString::from_vec(&["WhatCallIt", "20"]),
        CatalogItemString::from_vec(&["ThingaMaJig", "30"]),
    ]);
    glue.given_shopping_cart(&[ShoppingCartString { items: "=EmptyCart".to_string(), shipping: "$0".to_string(), discount: "$0".to_string(), totalprice: "$0".to_string(), shippingaddress: AddressString { street: "".to_string(), city: "".to_string(), state: "".to_string(), zip: "".to_string() }, billingaddress: AddressString { street: "".to_string(), city: "".to_string(), state: "".to_string(), zip: "".to_string() } }]);
    glue.when_item_added(&[OrderItemString::from_vec(&["Widget", "2", "1", "1"])]);
    glue.when_item_added(&[OrderItemString::from_vec(&["WhatCallIt", "3", "1", "1"])]);
    glue.then_shopping_cart_is(&[ShoppingCartString { items: "=TwoItemCart".to_string(), shipping: "$5".to_string(), discount: "$4".to_string(), totalprice: "$81".to_string(), shippingaddress: AddressString { street: "".to_string(), city: "".to_string(), state: "".to_string(), zip: "".to_string() }, billingaddress: AddressString { street: "".to_string(), city: "".to_string(), state: "".to_string(), zip: "".to_string() } }]);
}

#[test]
fn scenario_cost_of_empty_orderitemcollection() {
    let mut glue = ShoppingCartGlue::new();
    glue.given_catalog_has(&[
        CatalogItemString::from_vec(&["Widget", "10"]),
        CatalogItemString::from_vec(&["WhatCallIt", "20"]),
        CatalogItemString::from_vec(&["ThingaMaJig", "30"]),
    ]);
    glue.given_item_collection(&[
    ]);
    glue.then_total_of_items_is(&[ItemPriceInputString::from_vec(&["$0"])]);
}

// --- BusinessRule Tests ---

#[test]
fn business_rule_total_cart_price() {
    let mut glue = ShoppingCartGlue::new();
    glue.examples_business_rule_total_cart_price(&[
        CartInputString::from_vec(&["$110", "$5", "$11", "$104", "Discount applied before shipping calculated"]),
        CartInputString::from_vec(&["$80", "$5", "$4", "$81", ""]),
    ]);
}

#[test]
fn business_rule_shipping_cost() {
    let mut glue = ShoppingCartGlue::new();
    glue.examples_business_rule_shipping_cost(&[
        ShippingInputString::from_vec(&["$99.99", "$5.00", "Less than $100"]),
        ShippingInputString::from_vec(&["$100.00", "$0", "Free if $100 or more"]),
    ]);
}

#[test]
fn business_rule_discount() {
    let mut glue = ShoppingCartGlue::new();
    glue.examples_business_rule_discount(&[
        DiscountInputString::from_vec(&["$24.99", "0", ""]),
        DiscountInputString::from_vec(&["$25.00", "5", ""]),
        DiscountInputString::from_vec(&["$99.99", "5", ""]),
        DiscountInputString::from_vec(&["$100.00", "10", ""]),
    ]);
}

// --- DataType Tests ---

#[test]
fn data_type_percentage() {
    let mut glue = ShoppingCartGlue::new();
    glue.examples_data_type_percentage(&[
        ValidValuesString::from_vec(&["0", "y", ""]),
        ValidValuesString::from_vec(&["99", "y", ""]),
        ValidValuesString::from_vec(&["100", "y", ""]),
        ValidValuesString::from_vec(&["101", "n", ""]),
        ValidValuesString::from_vec(&["-1", "n", ""]),
    ]);
}

